cpp/Platform.Delegates/Platform.Delegates.MulticastDelegate.h

```diff
@@ -126,57 +88,98 @@ namespace Platform::Delegates
 
         MulticastDelegate &operator+=(const DelegateType &callback)
         {
-            const std::lock_guard<std::mutex> lock(mutex);
+            std::lock_guard lock(mutex);
             this->callbacks.emplace_back(callback);
             return *this;
         }
 
-        MulticastDelegate &operator+=(DelegateRawFunctionType &callback)
+        MulticastDelegate &operator+=(DelegateRawFunctionType callback)
         {
-            return *this += DelegateType(callback);
+            return *this += DelegateType{callback};
         }
 
         MulticastDelegate &operator+=(const DelegateFunctionType &callback)
         {
-            return *this += DelegateType(callback);
+            return *this += DelegateType{callback};
         }
 
         MulticastDelegate &operator-=(const DelegateType &callback)
         {
-            const std::lock_guard<std::mutex> lock(mutex);
+            const std::lock_guard lock{mutex};
             auto searchResult = std::find(this->callbacks.rbegin(), this->callbacks.rend(), callback);
             if (searchResult != this->callbacks.rend()) {
-                auto removedCallbackPosition = --(searchResult.base());
-                this->callbacks.erase(removedCallbackPosition);
+                this->callbacks.erase(searchResult.base());
             }
             return *this;
         }
 
         MulticastDelegate &operator-=(DelegateRawFunctionType &callback)
         {
-            return *this -= DelegateType(callback);
+            return *this -= DelegateType{callback};
         }
 
         MulticastDelegate &operator-=(const DelegateFunctionType &callback)
         {
-            return *this -= DelegateType(callback);
+            return *this -= DelegateType{callback};
         }
 
-        virtual ReturnType operator()(Args... args) override
+        ReturnType operator()(Args... args) override
         {
-            const std::lock_guard<std::mutex> lock(mutex);
-            auto callbacksSize = this->callbacks.size();
-            if (callbacksSize == 0)
+            const std::lock_guard lock(mutex);
+            if (this->callbacks.size() == 0)
             {
                 throw std::bad_function_call();
             }
-            auto lastElement = callbacksSize - 1;
-            for (size_t i = 0; i < lastElement; i++)
+            for (auto callbackIt = this->callbacks.rbegin();
+                    callbackIt != std::prev(this->callbacks.rend());
+                    ++callbackIt)
             {
-                this->callbacks[i](std::forward<Args>(args)...);
+                (*callbackIt)(args...);
```

Comment:
Are you sure it will work ok?

Reply:
For an iterator, this will definitely work just as well. I removed `std::forward` because it is only needed when using universal references, details can be found here -> https://isocpp.org/blog/2012/11/universal-references-in-c11-scott-meyers

Reply:
Well.. Ok.